### ISSUE TITLE
On-demand const verification may be a backwards compatibility hazard

### ISSUE THREAD

**Issue by huonw** [rust-lang/rust#19265]
E.g. #19244 describes how the following is valid, but uncommenting the line causes the error given below.

``` rust
struct Wrapper(usize);
const BAD_CONST: usize = Wrapper(42).0;

// type X = [u8; BAD_CONST];

fn main() {}
```

```rust
3 | const BAD_CONST: usize = Wrapper(42).0;
  |                          ^^^^^^^^^^^ unimplemented constant expression: tuple struct constructors
```

(NB. that specific instance may be fixed in future, but that does not necessarily mean this bug is fixed.)

That is, the const-suitability of the value is checked when used as a constant expression, not at the definition. This could possibly may lead to surprises if rules are tweaked (at least, it could lead to surprises that are not detected). In any case, it is certainly confusing for a `const` to apparently be an unsupported constant expression.

It seems that we could check the definitions, not just checking them lazily when they are used as const expressions.

**Comment by oli-obk** [comment]
Related issue: https://github.com/rust-lang/rust/issues/3170#issuecomment-158410569

```rust
const ARR: [u32; 0] = []; const X: u32 = ARR[5];
```

The above doesn't error until you do `let x = X;` somewhere.

**Comment by eddyb** [comment]
I've updated the issue description so that it matches what's happening today.
*However*, the only reason that example still behaves the way it does is [this whitelist](https://github.com/rust-lang/rust/blob/c67cf5fddc1d2807e9942c1f9793c3230367332b/src/librustc_passes/consts.rs#L70) - which is somewhat necessary as that expression is *quite valid* and will work when put through the trans evaluator (or miri).

@oli-obk's example is more interesting and I'm not sure how that one is getting whitelisted.

**Comment by oli-obk** [comment]
This has been solved. All exported constants are checked even if unused
